api/client/gro_client.py

```diff
@@ -370,22 +370,37 @@ def find_data_series(self, **kwargs):
             search_results.append(
                 self.search('regions', kwargs['partner_region'])[:cfg.MAX_RESULT_COMBINATION_DEPTH])
             keys.append('partner_region_id')
-        all_data_series = []
+        # Rank by frequency and source, while preserving search ranking in
+        # permutations of item, metric, region, and partner region.
+        freq_ranking = []
+        ranking_groups = set()
+        count = 0
         for comb in itertools.product(*search_results):
             entities = dict(list(zip(keys, [entity['id'] for entity in comb])))
-            data_series_list = self.get_data_series(**entities)
-            self._logger.debug("Found {} distinct data series for {}".format(
-                len(data_series_list), entities))
-            # temporal coverage affects ranking so add time range if specified.
-            for data_series in data_series_list:
-                if kwargs.get('start_date'):
-                    data_series['start_date'] = kwargs['start_date']
-                if kwargs.get('end_date'):
-                    data_series['end_date'] = kwargs['end_date']
-            all_data_series += data_series_list
-        self._logger.info("Found {} distinct data series total for {}".format(
-            len(all_data_series), kwargs))
-        for data_series in self.rank_series_by_source(all_data_series):
+            for data_series in self.get_data_series(**entities)[:cfg.MAX_SERIES_PER_COMB]:
+                count += 1
+                self._logger.debug("Data series: {}".format(data_series))
+                # time range affects ranking
+                data_series.pop('start_date', None)
```

Comment:
Doing the inverse seems like it'd be safer here:

```py
data_series = {key: data_series[key] for key in data_series.keys()
               if key in ['metric_id', 'item_id', 'region_id', 'partner_region_id']}
```

just in case we added something to the data_series objects in the future that might break the `frozenset` hashing (like how adding metadata and it being an object messed up pandas hashing before).

Reply:
yeah I was hesitating about this but concluded it is more understandable this way. What we're doing is not saying we only want item/metric/region/partner_region  or just all series selection entities like ENTITY_KEY_TO_TYPE above. We are saying we specifically want to remove the things we're going to rank on i.e. time and freq, then source. so I decided to go with explicit removal than explicit addition. 

THere are more comments in the subsequent pr
https://github.com/gro-intelligence/api-client/pull/192/files#diff-603e79178b8e522d0e366a9fca477b24R388